File: runtime/cpp/tests/zenoh_service_client.cpp
/// 跨进程 smoke：C++ zenoh service client 端。
///
/// 用法：./zenoh_service_client
/// 环境变量 FLOWRT_ZENOH_SERVICE_NAME 可覆盖 service name。

#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <flowrt/runtime.hpp>
#include <string>

struct AddRequest {
    std::int32_t a{};
    std::int32_t b{};

    static constexpr std::size_t wire_size() noexcept {
        return sizeof(std::int32_t) + sizeof(std::int32_t);
    }

    void encode_wire(std::span<std::uint8_t> output) const {
        flowrt::ensure_wire_size(wire_size(), output.size());
        flowrt::write_wire_le(output, 0, a);
        flowrt::write_wire_le(output, sizeof(std::int32_t), b);
    }

    static AddRequest decode_wire(std::span<const std::uint8_t> input) {
        flowrt::ensure_wire_size(wire_size(), input.size());
        return AddRequest{flowrt::read_wire_le<std::int32_t>(input, 0),
                          flowrt::read_wire_le<std::int32_t>(input, sizeof(std::int32_t))};
    }
};

struct AddResponse {
    std::int32_t sum{};

    static constexpr std::size_t wire_size() noexcept { return sizeof(std::int32_t); }

    void encode_wire(std::span<std::uint8_t> output) const {
        flowrt::ensure_wire_size(wire_size(), output.size());
        flowrt::write_wire_le(output, 0, sum);
    }

    static AddResponse decode_wire(std::span<const std::uint8_t> input) {
        flowrt::ensure_wire_size(wire_size(), input.size());
        return AddResponse{flowrt::read_wire_le<std::int32_t>(input, 0)};
    }
};

int main() {
    const char *name_env = std::getenv("FLOWRT_ZENOH_SERVICE_NAME");
    std::string name = name_env ? name_env : "flowrt/cross_lang/add";

    auto session = std::make_shared<::zenoh::Session>(flowrt::zenoh::open_zenoh_session_from_env());

    auto client = flowrt::zenoh::ZenohServiceClient<AddRequest, AddResponse>::open(name, session);

    std::fprintf(stderr, "[cpp-client] calling service '%s'...\n", name.c_str());

    auto result = client.call(AddRequest{10, 20}, 5000);
    if (result.is_ok()) {
        std::fprintf(stderr, "[cpp-client] OK: 10 + 20 = %d\n", result.value()->sum);
        if (result.value()->sum != 30) {
            std::fprintf(stderr, "[cpp-client] FAIL: expected 30, got %d\n", result.value()->sum);
            return 1;
        }
        std::fprintf(stderr, "[cpp-client] PASS\n");
        return 0;
    } else {
        std::fprintf(stderr, "[cpp-client] FAIL: code=%d msg=%s\n",
                     static_cast<int>(result.error_code()),
                     result.error_message().value_or("").c_str());
        return 1;
    }
}

File: runtime/rust/examples/zenoh_service_client.rs
//! 或编译为独立 binary 运行。

#![cfg(feature = "zenoh")]

use flowrt::{
    ServiceResult, WireCodec, WireCodecError,
    zenoh::{ZenohServiceClient, config_from_environment},
};
use zenoh::Wait;

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AddRequest {
    pub a: i32,
    pub b: i32,
}

impl WireCodec for AddRequest {
    const WIRE_SIZE: usize = 8;
    fn encode_wire(&self, output: &mut [u8]) -> Result<(), WireCodecError> {
        if output.len() != Self::WIRE_SIZE {
            return Err(WireCodecError::wrong_size(Self::WIRE_SIZE, output.len()));
        }
        output[..4].copy_from_slice(&self.a.to_le_bytes());
        output[4..].copy_from_slice(&self.b.to_le_bytes());
        Ok(())
    }
    fn decode_wire(input: &[u8]) -> Result<Self, WireCodecError> {
        if input.len() != Self::WIRE_SIZE {
            return Err(WireCodecError::wrong_size(Self::WIRE_SIZE, input.len()));
        }
        Ok(Self {
            a: i32::from_le_bytes([input[0], input[1], input[2], input[3]]),
            b: i32::from_le_bytes([input[4], input[5], input[6], input[7]]),
        })
    }
}

#[repr(C)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct AddResponse {
    pub sum: i32,
}

impl WireCodec for AddResponse {
    const WIRE_SIZE: usize = 4;
    fn encode_wire(&self, output: &mut [u8]) -> Result<(), WireCodecError> {
        if output.len() != Self::WIRE_SIZE {
            return Err(WireCodecError::wrong_size(Self::WIRE_SIZE, output.len()));
        }
        output.copy_from_slice(&self.sum.to_le_bytes());
        Ok(())
    }
    fn decode_wire(input: &[u8]) -> Result<Self, WireCodecError> {
        if input.len() != Self::WIRE_SIZE {
            return Err(WireCodecError::wrong_size(Self::WIRE_SIZE, input.len()));
        }
        Ok(Self {
            sum: i32::from_le_bytes([input[0], input[1], input[2], input[3]]),
        })
    }
}

pub fn service_name() -> String {
    std::env::var("FLOWRT_ZENOH_SERVICE_NAME")
        .unwrap_or_else(|_| "flowrt/cross_lang/add".to_string())
}

pub fn open_session() -> zenoh::Session {
    zenoh::open(config_from_environment().unwrap_or_default())
        .wait()
        .expect("zenoh session should open")
}

fn main() {
    let session = open_session();
    let name = service_name();

    let client = ZenohServiceClient::<AddRequest, AddResponse>::open(&name, session.clone());

    eprintln!("[rust-client] calling service '{name}'...");

    let result = client.call(AddRequest { a: 10, b: 20 }, 5000);
    match result {
        ServiceResult::Ok(resp) => {
            eprintln!("[rust-client] OK: 10 + 20 = {}", resp.sum);
            assert_eq!(resp.sum, 30, "expected 30, got {}", resp.sum);
            eprintln!("[rust-client] PASS");
        }
        ServiceResult::Err(code, msg) => {
            eprintln!("[rust-client] FAIL: code={code:?} msg={msg:?}");
            std::process::exit(1);
        }
    }
}
